python_modules/dagster/dagster/_core/definitions/asset_daemon_cursor.py

```diff
@@ -11,52 +13,44 @@
     TypeVar,
 )
 
-import dagster._check as check
+from dagster._core.definitions.asset_graph_subset import AssetGraphSubset
+from dagster._core.definitions.asset_subset import AssetSubset
+from dagster._core.definitions.auto_materialize_rule_evaluation import (
+    BackcompatAutoMaterializeAssetEvaluationSerializer,
+)
 from dagster._core.definitions.events import AssetKey
-from dagster._core.definitions.time_window_partitions import (
-    TimeWindowPartitionsDefinition,
-    TimeWindowPartitionsSubset,
+from dagster._core.definitions.partition import PartitionsDefinition
+from dagster._serdes.serdes import (
+    _WHITELIST_MAP,
+    PackableValue,
+    WhitelistMap,
+    deserialize_value,
+    serialize_value,
+    whitelist_for_serdes,
 )
-from dagster._serdes.serdes import PackableValue, deserialize_value, serialize_value
 
 from .asset_graph import AssetGraph
-from .asset_subset import AssetSubset
-from .partition import PartitionsDefinition, PartitionsSubset
 
 if TYPE_CHECKING:
     from .asset_condition import AssetCondition, AssetConditionEvaluation, AssetConditionSnapshot
-    from .asset_condition_evaluation_context import AssetConditionEvaluationContext
-
-ExtrasDict = Mapping[str, PackableValue]
 
 T = TypeVar("T")
 
 
-def _get_placeholder_missing_condition() -> "AssetCondition":
-    """Temporary hard-coding of the hash of the "materialize on missing" condition. This will
-    no longer be necessary once we start serializing the AssetDaemonCursor.
-    """
-    from .asset_condition import RuleCondition
-    from .auto_materialize_rule import MaterializeOnMissingRule
-
-    return RuleCondition(MaterializeOnMissingRule())
-
-
-_PLACEHOLDER_HANDLED_SUBSET_KEY = "handled_subset"
-
-
+@whitelist_for_serdes
 class AssetConditionCursorExtras(NamedTuple):
-    """Class to represent additional unstructured information that may be tracked by a particular
-    asset condition.
+    """Represents additional state that may be optionally saved by an AssetCondition between
+    evaluations.
     """
 
     condition_snapshot: "AssetConditionSnapshot"
-    extras: ExtrasDict
+    extras: Mapping[str, PackableValue]
```

Comment:
Have you thought about how "extras" will play out as a user-facing name? What are the APIs that we'd expose where users would interact with it? Might it end up in the UI?

Reply:
good question -- for the built-in rules, I wasn't imagining this ever making its way into the UI, as it's really just an implementation detail. but there's definitely a world in which users might need to work with this for custom rules, in which case it might end up in the UI for debugging purposes (although I don't think that's a hard requirement)

Purely on the python level, I'd expect users to interact with it through the context something like: 

```python

def my_eval_fn(context):
    subset_with_some_property = context.extras
    subset_with_some_property |= ...
    context.set_extras(subset_with_some_property)
    ...
```

This pattern is pretty similar to the existing cursor pattern we have for sensors, but this is more permissive in what's allowed to be stored in this field (i.e. instead of just a raw string, you can put any serializable object, which is a pretty nice QoL improvement)

Reply:
What would the docstring for set_extras look like?

Reply:
maybe something along the lines of:

```python
# method on the user context object
def update_extra_value(self, value: PackableValue) -> None:
    """Updates the extra value for this asset condition, which will be provided on the context for the
        next condition evaluation.

        This can be used to keep track of progress and avoid duplicate work across sensor
        evaluations.
        """

```

I'm of two minds on if this should actually just be called "cursor" in user-facing land (due to its similarities with the existing sensor cursor concept, and the fact that they likely won't need to / be allowed to interact directly with the "real" cursor object)

We do have a fair amount of latitude in choosing whatever user-facing name we want for this (although it would be best if the names lined up of course), but in a world where we showed "cursor" to the user, I'd probably still want to call these values "extra_values" or something of the sort in our internal storage representation (as they really are extra values on top of some other important cursor information)

Reply:
Is "cursor" definitely the right name for the non-extra data?

Reply:
I think "cursor" pretty accurately describes the purpose of the data that we're tracking here. We could consider calling it something like "previous evaluation state", but I don't think that really frees up the word "cursor" for the purely-extra data.

Maybe we could just call the "extra" data "extra_state"? The user can then grab "extra_state" off the context and update "extra_state" as desired.

Reply:
https://github.com/dagster-io/dagster/pull/19035/files this PR updates those names, so we can hash out the final naming there